controllers/clustercache/cluster_cache.go

```diff
@@ -0,0 +1,650 @@
+		return nil, errors.Wrapf(ErrClusterNotConnected, "failed to get client")
+	}
+	return accessor.GetClient(ctx)
+}
+
+func (cc *clusterCache) GetReader(ctx context.Context, cluster client.ObjectKey) (client.Reader, error) {
+	accessor := cc.getClusterAccessor(ctx, cluster)
+	if accessor == nil {
+		return nil, errors.Wrapf(ErrClusterNotConnected, "failed to get client reader")
+	}
+	return accessor.GetReader(ctx)
+}
+
+func (cc *clusterCache) GetRESTConfig(ctx context.Context, cluster client.ObjectKey) (*rest.Config, error) {
+	accessor := cc.getClusterAccessor(ctx, cluster)
+	if accessor == nil {
+		return nil, errors.Wrapf(ErrClusterNotConnected, "failed to get rest config")
+	}
+	return accessor.GetRESTConfig(ctx)
+}
+
+func (cc *clusterCache) GetClientCertificatePrivateKey(ctx context.Context, cluster client.ObjectKey) (*rsa.PrivateKey, error) {
+	accessor := cc.getClusterAccessor(ctx, cluster)
+	if accessor == nil {
+		return nil, errors.New("failed to get client certificate private key: private key was not generated yet")
+	}
+	return accessor.GetClientCertificatePrivateKey(ctx), nil
+}
+
+func (cc *clusterCache) Watch(ctx context.Context, input WatchInput) error {
+	accessor := cc.getClusterAccessor(ctx, input.Cluster)
+	if accessor == nil {
+		return errors.Wrapf(ErrClusterNotConnected, "error creating watch %s for %T on Cluster %s", input.Name, input.Kind, input.Cluster.String())
+	}
+	return accessor.Watch(ctx, input)
+}
+
+func (cc *clusterCache) GetLastProbeSuccessTimestamp(ctx context.Context, cluster client.ObjectKey) time.Time {
+	accessor := cc.getClusterAccessor(ctx, cluster)
+	if accessor == nil {
+		return time.Time{}
+	}
+	return accessor.GetLastProbeSuccessTimestamp(ctx)
+}
+
+const (
+	// defaultRequeueAfter is used as a fallback if no specific requeue should be used.
+	defaultRequeueAfter = 10 * time.Second
+)
+
+// Reconcile reconciles Clusters and manages corresponding clusterAccessors.
+func (cc *clusterCache) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	clusterKey := client.ObjectKey{Namespace: req.Namespace, Name: req.Name}
+
+	accessor := cc.getOrCreateClusterAccessor(ctx, clusterKey)
+
+	cluster := &clusterv1.Cluster{}
+	if err := cc.client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(4).Info("Cluster has been deleted, disconnecting cluster accessor")
+			accessor.Disconnect(ctx)
+			cc.deleteClusterAccessor(ctx, clusterKey)
+			return ctrl.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		log.Error(err, "Failed to read the Cluster object")
+		return ctrl.Result{RequeueAfter: defaultRequeueAfter}, nil
+	}
+
+	// Return if infrastructure is not ready yet to avoid trying to open a connection when it cannot succeed.
+	// Requeue is not needed as there will be a new reconcile.Request when Cluster.status.infrastructureReady is set.
+	if !cluster.Status.InfrastructureReady {
```

Comment:
I think then we're deadlocked. IIRC we need access to the workload cluster to even get to the point where we have a controlPlaneEndpoint

Reply:
I don't know if it's safe to wait for that or if that means we're hitting deadlocks with some providers.

What I do know is that it was not safe to wait until KCP was initialized, because KCP has to access the wl cluster via the ClusterCache to set initialized to true (it checks for the kubeadm ConfigMap in the wl cluster).
(I know it's a slightly orthogonal example, but just to show what sort of dependencies there might exist)

So I'm a bit hesitant to add this to avoid introducing regressions.